src/matrix_auth_observer.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Matrix Auth relation observer."""
+
+import logging
+from collections import namedtuple
+from pathlib import Path
+from typing import List, Optional
+
+import ops
+from charms.synapse.v0.matrix_auth import (
+    MatrixAuthProviderData,
+    MatrixAuthProvides,
+    MatrixAuthRequirerData,
+)
+from ops.framework import Object
+
+import synapse
+from charm_state import CharmBaseWithState, CharmState, inject_charm_state
+
+logger = logging.getLogger(__name__)
+
+
+class MatrixAuthObserver(Object):
+    """The Matrix Auth relation observer."""
+
+    def __init__(self, charm: CharmBaseWithState):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "matrix-auth-observer")
+        self._charm = charm
+        self.matrix_auth = MatrixAuthProvides(self._charm)
+        # matrix_auth_request_received conflicts with on defined by Redis...
+        self.framework.observe(
+            self._charm.on["matrix-auth"].relation_changed, self._on_matrix_auth_relation_changed
+        )
+        self.framework.observe(
+            self._charm.on["matrix-auth"].relation_departed, self._on_matrix_auth_relation_departed
+        )
+
+    def get_charm(self) -> CharmBaseWithState:
+        """Return the current charm.
+
+        Returns:
+           The current charm
+        """
+        return self._charm
+
+    def update_matrix_auth_integration(self, charm_state: CharmState) -> None:
+        """Update matrix auth integration relation data.
+
+        Args:
+            charm_state: The charm state.
+        """
+        for relation in list(self._charm.model.relations["matrix-auth"]):
+            if not relation.units:
+                return
+            provider_data = self._get_matrix_auth_provider_data(charm_state)
+            if self._matrix_auth_relation_updated(relation, provider_data):
+                return
+            self.matrix_auth.update_relation_data(relation, provider_data)
+
+    def get_requirer_registration_secrets(self) -> Optional[List]:
+        """Get requirers registration secrets (application services).
+
+        Returns:
+            dict with filepath and content for creating the secret files.
+        """
+        registration_secrets = []
+        RegistrationSecret = namedtuple("RegistrationSecret", ["file_path", "value"])
```

Comment:
Would it help to have a separate class defined for it?
```suggestion
class RegistrationSecret(typing.NamedTuple):
    file_path: pathlib.Path
    value: str
```